### ISSUE TITLE
Don't initialize variable to `undefined`.

### ISSUE THREAD

**Issue by leewz** [google/closure-compiler#2804]
(Using https://closure-compiler.appspot.com/home with Advanced Optimizations.)

The following

    var a = undefined;
    try {
      mayThrow();
    } catch(b) {
      a = 0;
    }
    console.log(a);

is compiled to (with Pretty Print on for readability)

    var a = void 0;
    try {
      mayThrow();
    } catch (b) {
      a = 0;
    }
    console.log(a);

If the `undefined` is removed from the first line, the compiler doesn't initialize the variable `a`.

Is there a use for explicitly initializing a variable to `undefined`? It takes at least seven more characters (`=void 0`).

**Comment by lauraharker** [comment]
Replacing every `let a = undefined` with `let a;` as a peephole optimization seems safe. (Except if someone shadows `undefined`, but the compiler might already assume that never happens).

`var a  = undefined;` is trickier because of hoisting. e.g.
```
a = 4;
console.log(a); // prints 4
var a = undefined; // removing the "= undefined" would change the next statement's output
console.log(a); // prints undefined
```
I'm sure it's possible to handle this case safely, just more complicated...

**Comment by brad4d** [comment]
re: 1
The compiler does give warnings if you declare the same variable multiple times with `var`, though there are combinations of options that turn the warning off.

re: the rest
This falls into the large category of optimizations we could do, but don't do currently.
While this sounds pretty simple to implement, there's still a maintenance cost to it.
The amount of code size improvement looks to be minimal and unlikely to be worth that maintenance cost.
